### ISSUE TITLE
Add `to-principal` to convert a <trait> to a `principal`

### ISSUE THREAD

**Issue by psq** [blockstack/stacks-blockchain#1663]
To summarize what we have available so far:
- traits can can called, but not stored
- contract principals can be stored, compared, used as keys, but not called
- a contract principal and a trait both use an address and contract name

Proposal: implement `to-principal`, which takes a trait as a parameter, and returns a `principal`.

Benefits:
- simple implementation (I think...)
- does not require changing any static analysis code
- does not affect language decidability properties
- allows for white listing traits
- allows for storing data/metadata on passed in traits
- other use cases?

Some background:

To implement something like Moloch (see [Support storing traits in variables](https://github.com/blockstack/stacks-blockchain/issues/1659) submitted by @friedger), we would need to store the principal of a token, and later be able to call it.  I think now everyone is in agreement that we don't want to allow this directly, so we can know ahead of time how much executing such a contract will cost.  A possible alternative is to pass the trait instead when there is a need to call the token further down the invocation chain, but in the case of Moloch, the contract can not just blindly call the trait, as it may not be the trait of a token that has been agreed upon for a given deal.

So, the contract needs to be able to check against a list of approved tokens (a whitelist).  If we could convert the trait to its contract, we could implement such a white list.  The trait can still be called, but not stored, but now that we have the contract principal we can look it up, store it for later retrieval, therefore we keep the basic principle intact and we are not affecting the decidability of the contract.

An other example involves `swapr`, since I could not store traits, or not use them as keys in a map, I started implementing a [Registry](https://github.com/psq/swapr/blob/master/contracts/swapr-registry.clar), but even though technically this would work, there are a number of issues with that solution.  One being that as anyone can add something to the registry, the contract added may not implement the expected methods, and even though it will fail when trying to invoke, this has polluted the registry, and the number of slots is limited.  Adding the ability to remove a pair would break the full `trustless` aspect.  I can not pass the trait, as it can't be stored, looked up or compared, and even if I passed both the contract principal and the trait, there is no way to check that these point to the same contract.

The second issue is that the contract can be malicious, and yes, some people may be able to review the contract before sending tokens to it, I would think few could or will.  And the blame will fall on `swapr`, not on whoever deployed and registered that contract.  And yes, there are other options I've looked into (manual review of the new contract, using a deployment service that holds the private key so the registry can be locked down to a specific `tx-sender`, and a few others).  But this breaks the 'fully trustless' aspect.  So not so viable in practice.  But if `to-principal` is available, then I would not need to deploy a new contract for each pair, instead I can have a map with the 2 token principals as key, and store things like balances, ownership, etc.

So... let me know what you think, whether the use cases enabled make sense, whether this is something that could be added to the Clarity VM, or whether I missed any critical downside.  @friedger, would that allow you to implement what you had in mind?

In the meantime, I'll start on a prototype implementation to make sure this can deliver on what I think we need.

Ping @friedger, @jcnelson, @lgalabru for comments, should I include anyone else?

**Comment by jcnelson** [comment]
Also @kantai

**Comment by jcnelson** [comment]
To answer the topic of this issue, I would welcome the addition of a function like `to-principal` if there's not a way to do it in Clarity already (but maybe `trait-principal` would be a more appropriate name, since it would only operate on trait implementations?).

To address your specific design rationale on the need for `to-principal`:

> the contract can not just blindly call the trait, as it may not be the trait of a token that has been agreed upon for a given deal.

I'm not sure I follow the reasoning here.  I think you're saying that you need a whitelist of token names paired with their trait principals, which would get used to authenticate the trait implementation that a user includes when buying or selling a token?  If so, then if anyone can register a token with a trait name in the whitelist, what stops a scammer from front-running your whitelist-token transaction and claiming the token for themselves?  If you wanted to register token `foo` with your trait implementation, then I could snipe your `foo` registration by watching the mempool for such transactions and submitting a higher-fee transaction to grab it first.

I'm not very familiar with Moloch, so please forgive me if this alternative proposal overlooks something obvious.  Suppose `swapr` defined two traits:  a "buyer" trait and a "seller" trait.  The buyer trait implementation would store tokens used to purchase other tokens, and would include a "check-seller" method that verified that the buyer received the desired token before executing the trade (so no one could sell them the wrong token).  Similarly, the seller trait would include a "check-buyer" method that verified that the seller received the payment in the right tokens (so the buyer wouldn't be able to steal the tokens).  `swapr` would then take two traits implementations as arguments -- the buyer and seller contracts -- and call the "check-buyer" and "check seller" methods before executing the trade methods.  As a user, I would then use `swapr` by implementing and deploying "buyer" and "seller"-compatible contracts, filling them up with tokens for buying and selling, and calling `swapr`'s public functions to execute and track trades.  Since the users' trait implementations take care of verifying that the right tokens were transferred, does this obviate the need for `to-principal`?

**Comment by psq** [comment]
Thank you for your thoughts, @jcnelson.

> maybe trait-principal would be a more appropriate name, since it would only operate on trait implementations?

that name would certainly work.

The solution you suggest, using a buyer trait and a seller trait would work in a case where you can implement what you need in a single transaction, and it would require a sophisticated user to be able to implement the traits, so this would also reduce the audience.

So let me try to simplify the use cases for `moloch` and `swapr`, 2 different use cases, but they both could be implemented with `to-principal`/`trait-principal`.

First they both involve what I'd call long running transactions, but not to confuse with the Clarity transactions, I'll call these `operations`, meaning you will have at least 2 transactions, or more, over long periods of time, potentially executed by multiple users, who probably don't trust each other so rely on the contracts to check no one is misbehaving.

## moloch case
User 1 calls `setup`:
```
setup(trait1, ...)
```

User 2 calls `complete`:
```
complete(trait1, ...)
```

The VM will check that the first parameter implements `trait1`, and invocation will fail it the parameter passed to `setup` or `complete` does not.  However, and that's the critical point, something we can not do today: How can the contract ensure that the contract passed for `trait1` during `setup` is the __same__ as the contract passed to `trait1` during `complete`.  If we can not guarantee this, all sort of bad things could ensue.  Being able to extract the principal from the trait would allow you to do this.

[BTW, if you'd use the solution requested in #1659, you don't have this problem, as you are guaranteed to use the same trait in all future calls as you'd have stored it.]

## swapr case
Called once to create a pair:
```
create_pair(traitx, traity, ...)
```
Called many times by different users:
```
add(traitx, traity, x, y, ...)
```
Called many times by a different set of users, although some overlap possible with those calling `add`:
```
swap-exact-x-for-y(traitx, traity, x, ...)
```

In this case, not only would the `swapr` contract need to verify that the contracts are the same as during creation, otherwise there's a clear breach of trust, but the __other key__ thing is that `swapr` needs to store data related to that pair (which can not be stored with either trait, as potentially there is also a trust issue.  You want to be able to guarantee that during the life of the long running `operation`, the exact same token contracts are being passed (guaranteed immutable), not just something that could masquerade as such, but have a malicious `transfer` function for example.  And if you have the principals for both trait, you can use a map using the 2 principals as key to store and retrieve data related to the pair.

Again, being able to extract the principal from the trait allows you to implement this second use case.

Hopefully, this helps explain what I'm trying to do with `swapr`, and hopefully would help @friedger for what he needs for implementing something like Moloch.

**Comment by friedger** [comment]
Molch (https://github.com/MolochVentures/moloch) is a simple DAO. It maintains a balance of tokens. These tokens can be added by proposals that are voted on, hence can change dynamically.

This proposal helps to implement a function that adds supported tokens of the balance map
```
(define-map balances (tuple (user principal) (token principal)) (tuple (amount uint)))
```

**Comment by kantai** [comment]
Having this functionality makes sense -- perhaps the most descriptive name for it would be `get-contract` or `contract-of` -- since the idea is that it returns the contract principal that implements the supplied trait argument?

**Comment by lgalabru** [comment]
We even initially implemented this native function, and got rid of it - https://github.com/blockstack/stacks-blockchain/pull/1246/commits/c9f524a18b21fbd354bdd34a07f54a7ecd838d72
Will restore it if we reached consensus.

**Comment by psq** [comment]
@lgalabru last night I created my own version, which is my first rust code, so there may be things that are not optimal.

I had to add more code as there are new things needed, like cost for example, and it checks for exactly 1 argument instead of 1 or more, and requires to use of `impl-trait` for the traits you use, which I would deem safer.

And I also included some tests.

Haven't created the PR yet, but here's the commit https://github.com/psq/stacks-blockchain/commit/39591ee82ebfe7de70dd3979c8ee3a883391d7ef

Plus my editor is set to remove trailing white spaces, which is a setting I need for other projects, so it looks like I made more changes than I actually did.  If that's a problem, I could redo some of the edits to show only actual changes.
